snapcraft/internal/deltas/_deltas.py

```diff
@@ -0,0 +1,199 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2016 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+
+import logging
+import os
+import shutil
+import subprocess
+
+from snapcraft.internal.deltas.errors import (
+    DeltaGenerationError,
+    DeltaFormatIsNoneError,
+    DeltaToolPathIsNoneError,
+    DeltaFormatOptionError,
+)
+
+
+logger = logging.getLogger(__name__)
+
+
+delta_formats_options = [
+    'xdelta'
+]
+
+
+class BaseDeltasGenerator:
+    """Class for delta generation
+
+    This class is responsible for the snap delta file generation
+    """
+
+    delta_format = None
+    delta_file_extname = 'delta'
+    delta_tool_path = None
+
+    def __init__(self, source_path, target_path):
+        self.source_path = source_path
+        self.target_path = target_path
+        self.pre_check()
+
+    def pre_check(self):
+        self._check_properties()
+        self._check_file_existence()
+        self._check_delta_gen_tool()
+
+    def _check_properties(self):
+        if self.delta_format is None:
```

Comment:
can the messages here be part of the exception? Also user friendly, specially in the case of tool paths' as that is user visible.

Look at snapcraft.storeapi.errors or snapcraft.errors for inspiration or the ""dogmatic"" way to do it